### PR TITLE
Allow an API Publisher to create subscription

### PR DESCRIPTION
**Pull request by phiz71**
## Issue

https://gravitee.atlassian.net/browse/APIM-2023

## Description

This PR adds a dialog accessible from the subscriptions list screen that allow to create a subscription for Standard plans.

Another PR is needed to handle Push plans.

+

Little refactoring in the edit component.

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-topnovngad.chromatic.com)

### CODE REVIEW

File: gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/components/creation-dialog/api-portal-subscription-creation-dialog.component.html [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        arial-label="Select an plan"
+        class="subscription-creation__content__plans"
+      >
+        <mat-radio-button *ngFor="let plan of plans" [value]="plan" [disabled]="!!plan.generalConditions || plan.mode === 'PUSH'">
+          {{ plan.name }}
+        </mat-radio-button>
+      </mat-radio-group>
+
+      <mat-form-field class="subscription-creation__content__custom-apikey" *ngIf="shouldDisplayCustomApiKey()">
+        <mat-label>Custom API-Key</mat-label>
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/components/creation-dialog/api-portal-subscription-creation-dialog.component.html:59]: Just a note: API Key is spelled lots of different ways across components 😅 `API-Key` vs `API key` vs `apikey`. I can send a message to @anthonyjgatti for l8r

File: gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/components/creation-dialog/api-portal-subscription-creation-dialog.component.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+  public form: FormGroup = new FormGroup({
+    selectedPlan: new FormControl(undefined, [Validators.required]),
+    selectedApplication: new FormControl(undefined, [applicationSelectionRequiredValidator]),
+  });
+
+  private unsubscribe$: Subject<boolean> = new Subject<boolean>();
+
+  constructor(
+    private readonly dialogRef: MatDialogRef<ApiPortalSubscriptionCreationDialogComponent, ApiPortalSubscriptionCreationDialogResult>,
+    @Inject(MAT_DIALOG_DATA) dialogData: { plans: Plan[] },
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/components/creation-dialog/api-portal-subscription-creation-dialog.component.ts]: You don't want ApiPortalSubscriptionCreationDialogData here?

File: gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/components/creation-dialog/api-portal-subscription-creation-dialog.component.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  }
+
+  ngOnInit(): void {
+    this.showGeneralConditionsMsg = this.plans.some((plan) => plan.generalConditions);
+
+    this.applications$ = this.form.get('selectedApplication').valueChanges.pipe(
+      distinctUntilChanged(),
+      debounceTime(100),
+      switchMap((term) => {
+        return term.length > 0 ? this.applicationService.list('ACTIVE', term, 'name', 1, 20) : of(new PagedResult<Application>());
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/components/creation-dialog/api-portal-subscription-creation-dialog.component.ts]: can return directly

**Review by jourdiw — APPROVED** [review]

File: gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/components/creation-dialog/api-portal-subscription-creation-dialog.component.ts (hunk 2 above)
Comment (phiz71) [reply gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/components/creation-dialog/api-portal-subscription-creation-dialog.component.ts]: I will change this

File: gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/list/api-portal-subscription-list.component.spec.ts [hunk 4]
```diff
@@ -94,20 +109,25 @@ describe('ApiPortalSubscriptionListComponent', () => {
   describe('filters tests', () => {
     it('should display default filters', fakeAsync(async () => {
       await initComponent([]);
+      const harness = await loader.getHarness(ApiPortalSubscriptionListHarness);
 
-      const planSelectInput = await loader.getHarness(MatSelectHarness.with({ selector: '[formControlName="planIds"]' }));
+      const planSelectInput = await harness.getPlanSelectInput();
       expect(await planSelectInput.isDisabled()).toEqual(false);
       expect(await planSelectInput.isEmpty()).toEqual(true);
+      // const plansOptions = await planSelectInput.getOptions();
```
Comment (ThibaudAV) [inline gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/list/api-portal-subscription-list.component.spec.ts:117]: to remove ?

File: gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/list/api-portal-subscription-list.component.ts [hunk 5]
```diff
@@ -227,8 +235,33 @@ export class ApiPortalSubscriptionListComponent implements OnInit, OnDestroy {
     this.ajsState.go(`${this.routeBase}.subscription.edit`, { subscriptionId });
   }
 
-  public navigateToNewSubscription(): void {
-    this.ajsState.go(`${this.routeBase}.subscription.new`, { apiId: this.api.id });
+  public createSubscription(): void {
+    this.matDialog
+      .open<
+        ApiPortalSubscriptionCreationDialogComponent,
```
Comment (ThibaudAV) [inline gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/list/api-portal-subscription-list.component.ts:241]: 💌 good ts typing

File: gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/list/api-portal-subscription-list.component.spec.ts (hunk 4 above)
Comment (phiz71) [reply gravitee-apim-console-webui/src/management/api/portal/ng-subscriptions/list/api-portal-subscription-list.component.spec.ts:117]: Yes.
Or we wait my next PR

**Review by ThibaudAV — APPROVED** [review]

**Pull request merged**